-Astubs doesn't write type arguments for outer class when inner class present

When using `-Ainfer=stubs`, creating stub files for classes with inner classes fails to include type arguments in the outer class. Consider this class
```java
public class GenericClass<T> {
    public class InnerClass {
        // Create error to force generation of stub file, ignore.
        Object o = null;
    }
}
```
I compiled it with,
```bash
$CHECKERFRAMEWORK/checker/bin/javac -processor NullnessChecker -Ainfer=stubs GenericClass.java
```
When I inspect the generated stub file, I see
```java
import org.checkerframework.checker.initialization.qual.Initialized;
import org.checkerframework.checker.nullness.qual.Nullable;

class GenericClass {

  class InnerClass {

    // fields:

    @Initialized @Nullable java.lang.Object o;

  }
}
```
Notice that `GenericClass` has no type parameter.

When I try to use this stub file, I get an error:
```bash
$CHECKERFRAMEWORK/checker/bin/javac -processor NullnessChecker -Astubs=build GenericClass.java
```
```
warning: StubParser: annotateTypeParameters: mismatched sizes:  typeParameters (size 0)=[];  typeArguments (size 1)=[T extends Object];  decl=class GenericClass {      class InnerClass {          @Initialized         @Nullable         java.lang.Object o;     } };  elt=GenericClass (class com.sun.tools.javac.code.Symbol$ClassSymbol).; for more details, run with -AstubDebug
GenericClass.java:4: error: [assignment.type.incompatible] incompatible types in assignment.
        Object o = null;
                   ^
  found   : null
  required: @Initialized @NonNull Object
1 error
1 warning
```

Output for
```bash
$CHECKERFRAMEWORK/checker/bin/javac -processor NullnessChecker -Ainfer=stubs GenericClass.java -version -verbose -AprintAllQualifiers
```
[javac_create_stubs_output.txt](https://github.com/typetools/checker-framework/files/4881794/javac_create_stubs_output.txt)

Output for
```bash
$CHECKERFRAMEWORK/checker/bin/javac -processor NullnessChecker -Astubs=build GenericClass.java -version -verbose -AprintAllQualifiers
```

[javac_use_stubs_output.txt](https://github.com/typetools/checker-framework/files/4881795/javac_use_stubs_output.txt)

--- a/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
+++ b/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
@@ -453,10 +453,10 @@ public final class SceneToStubWriter {
             String basename, AClass aClass, PrintWriter printWriter) {
         String[] classNames = basename.split("\\$");
         TypeElement innermostTypeElt = aClass.getTypeElement();
-        TypeElement[] typeElements = getTypeElementsForClasses(innermostTypeElt, classNames);
         if (innermostTypeElt == null) {
             throw new BugInCF("typeElement was unexpectedly null in this aClass: " + aClass);
         }
+        TypeElement[] typeElements = getTypeElementsForClasses(innermostTypeElt, classNames);
 
         for (int i = 0; i < classNames.length; i++) {
             String nameToPrint = classNames[i];
@@ -469,8 +469,7 @@ public final class SceneToStubWriter {
             if (i == classNames.length - 1) {
                 // Only print class annotations on the innermost class, which corresponds to aClass.
                 // If there should be class annotations on another class, it will have its own stub
-                // file,
-                // which will eventually be merged with this one.
+                // file, which will eventually be merged with this one.
                 printWriter.print(formatAnnotations(aClass.getAnnotations()));
             }
             printWriter.print(nameToPrint);
@@ -496,9 +495,6 @@ public final class SceneToStubWriter {
      */
     private static TypeElement @SameLen("#2") [] getTypeElementsForClasses(
             TypeElement innermostTypeElt, String[] classNames) {
-        if (classNames.length == 1) {
-            return new TypeElement[] {innermostTypeElt};
-        }
         TypeElement[] result = new TypeElement[classNames.length];
         TypeElement elt = innermostTypeElt;
         for (int i = classNames.length - 1; i >= 0; i--) {

--- a/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
+++ b/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
@@ -20,12 +20,14 @@ import javax.lang.model.element.VariableElement;
 import javax.lang.model.type.ArrayType;
 import javax.lang.model.type.TypeKind;
 import javax.lang.model.type.TypeMirror;
+import org.checkerframework.checker.index.qual.SameLen;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.checkerframework.checker.signature.qual.BinaryName;
 import org.checkerframework.checker.signature.qual.DotSeparatedIdentifiers;
 import org.checkerframework.common.wholeprograminference.scenelib.ASceneWrapper;
 import org.checkerframework.javacutil.AnnotationUtils;
 import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.ElementUtils;
 import scenelib.annotations.Annotation;
 import scenelib.annotations.el.AClass;
 import scenelib.annotations.el.AField;
@@ -450,6 +452,11 @@ public final class SceneToStubWriter {
     private static int printClassDefinitions(
             String basename, AClass aClass, PrintWriter printWriter) {
         String[] classNames = basename.split("\\$");
+        TypeElement innermostTypeElt = aClass.getTypeElement();
+        TypeElement[] typeElements = getTypeElementsForClasses(innermostTypeElt, classNames);
+        if (innermostTypeElt == null) {
+            throw new BugInCF("typeElement was unexpectedly null in this aClass: " + aClass);
+        }
 
         for (int i = 0; i < classNames.length; i++) {
             String nameToPrint = classNames[i];
@@ -459,9 +466,15 @@ public final class SceneToStubWriter {
             } else {
                 printWriter.print("class ");
             }
-            printWriter.print(formatAnnotations(aClass.getAnnotations()));
+            if (i == classNames.length - 1) {
+                // Only print class annotations on the innermost class, which corresponds to aClass.
+                // If there should be class annotations on another class, it will have its own stub
+                // file,
+                // which will eventually be merged with this one.
+                printWriter.print(formatAnnotations(aClass.getAnnotations()));
+            }
             printWriter.print(nameToPrint);
-            printTypeParameters(aClass, printWriter);
+            printTypeParameters(typeElements[i], printWriter);
             printWriter.println(" {");
             if (aClass.isEnum(nameToPrint) && i != classNames.length - 1) {
                 // Print a blank set of enum constants if this is an outer enum.
@@ -470,6 +483,29 @@ public final class SceneToStubWriter {
             printWriter.println();
         }
         return classNames.length;
+    }
+
+    /**
+     * Uses getEnclosingElement to construct an array of TypeElements corresponding to the list of
+     * classes, starting from the innermost element.
+     *
+     * @param innermostTypeElt the innermost type element representing an inner class
+     * @param classNames the names of the containing classes, from outer to inner
+     * @return an array of TypeElements whose entry at a given index represents the type named at
+     *     that index in {@code classNames}
+     */
+    private static TypeElement @SameLen("#2") [] getTypeElementsForClasses(
+            TypeElement innermostTypeElt, String[] classNames) {
+        if (classNames.length == 1) {
+            return new TypeElement[] {innermostTypeElt};
+        }
+        TypeElement[] result = new TypeElement[classNames.length];
+        TypeElement elt = innermostTypeElt;
+        for (int i = classNames.length - 1; i >= 0; i--) {
+            result[i] = elt;
+            elt = ElementUtils.enclosingClass(elt);
+        }
+        return result;
     }
 
     /**
@@ -723,14 +759,10 @@ public final class SceneToStubWriter {
     /**
      * Prints the type parameters of the given class, enclosed in {@code <...>}.
      *
-     * @param aClass the class whose type parameters should be printed
+     * @param type the TypeElement representing the class whose type parameters should be printed
      * @param printWriter where to print the type parameters
      */
-    private static void printTypeParameters(AClass aClass, PrintWriter printWriter) {
-        TypeElement type = aClass.getTypeElement();
-        if (type == null) {
-            return;
-        }
+    private static void printTypeParameters(TypeElement type, PrintWriter printWriter) {
         List<? extends TypeParameterElement> typeParameters = type.getTypeParameters();
         printTypeParameters(typeParameters, printWriter);
     }

--- a/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
+++ b/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
@@ -14,6 +14,7 @@ import java.util.Map;
 import java.util.StringJoiner;
 import java.util.regex.Pattern;
 import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.Element;
 import javax.lang.model.element.TypeElement;
 import javax.lang.model.element.TypeParameterElement;
 import javax.lang.model.element.VariableElement;
@@ -24,10 +25,10 @@ import org.checkerframework.checker.index.qual.SameLen;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.checkerframework.checker.signature.qual.BinaryName;
 import org.checkerframework.checker.signature.qual.DotSeparatedIdentifiers;
+import org.checkerframework.common.value.qual.MinLen;
 import org.checkerframework.common.wholeprograminference.scenelib.ASceneWrapper;
 import org.checkerframework.javacutil.AnnotationUtils;
 import org.checkerframework.javacutil.BugInCF;
-import org.checkerframework.javacutil.ElementUtils;
 import scenelib.annotations.Annotation;
 import scenelib.annotations.el.AClass;
 import scenelib.annotations.el.AField;
@@ -494,12 +495,13 @@ public final class SceneToStubWriter {
      *     that index in {@code classNames}
      */
     private static TypeElement @SameLen("#2") [] getTypeElementsForClasses(
-            TypeElement innermostTypeElt, String[] classNames) {
+            TypeElement innermostTypeElt, String @MinLen(1) [] classNames) {
         TypeElement[] result = new TypeElement[classNames.length];
-        TypeElement elt = innermostTypeElt;
-        for (int i = classNames.length - 1; i >= 0; i--) {
-            result[i] = elt;
-            elt = ElementUtils.enclosingClass(elt);
+        result[classNames.length - 1] = innermostTypeElt;
+        Element elt = innermostTypeElt;
+        for (int i = classNames.length - 2; i >= 0; i--) {
+            elt = elt.getEnclosingElement();
+            result[i] = (TypeElement) elt;
         }
         return result;
     }

--- a/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
+++ b/framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java
@@ -486,10 +486,10 @@ public final class SceneToStubWriter {
     }
 
     /**
-     * Uses getEnclosingElement to construct an array of TypeElements corresponding to the list of
-     * classes, starting from the innermost element.
+     * Constructs an array of TypeElements corresponding to the list of classes.
      *
-     * @param innermostTypeElt the innermost type element representing an inner class
+     * @param innermostTypeElt the innermost type element: either an inner class or an outer class
+     *     without any inner classes that should be printed
      * @param classNames the names of the containing classes, from outer to inner
      * @return an array of TypeElements whose entry at a given index represents the type named at
      *     that index in {@code classNames}

--- a/framework/tests/whole-program-inference/non-annotated/OuterClassWithTypeParam.java
+++ b/framework/tests/whole-program-inference/non-annotated/OuterClassWithTypeParam.java
@@ -1,0 +1,9 @@
+// test file for https://github.com/typetools/checker-framework/issues/3438
+
+import testlib.wholeprograminference.qual.Sibling1;
+
+public class OuterClassWithTypeParam<T> {
+    public class InnerClass {
+        Object o = (@Sibling1 Object) null;
+    }
+}

